### ISSUE TITLE
Some tools expect vendor binaries to be under vendor/pkg

### ISSUE THREAD

**Issue by mem** [constabulary/gb#314]
Hi,

I ran into an issue with gocode today. I have a test project set up like this:

src/a/a.go:
```go
package main

import "b"

func main() {
        b.
}
```

src/vendor/b/b.go:
```go
package b

var B = 1

func Get() int {
        return 2
}
```

(the weird "b." line is to demonstrate the autocompletion issue with gocode)

I start the gocode server like this:

    $ gocode -s -sock unix -addr localhost:37373 -debug

and run the client like this:

    $ gocode -in src/a/a.go autocomplete 43
    Nothing to complete.

The server outputs is:

```
2015/08/20 19:21:32 Got autocompletion request for '/home/magallma/devel/go/test/src/a/a.go'
2015/08/20 19:21:32 Cursor at: 43
2015/08/20 19:21:32 -------------------------------------------------------
package main

import "b"

func main() {
        b.#
}
2015/08/20 19:21:32 -------------------------------------------------------
2015/08/20 19:21:32 Import path "b" was not resolved
2015/08/20 19:21:32 Gocode's build context is:
2015/08/20 19:21:32  GOROOT: /home/magallma/devel/go/go-1.5
2015/08/20 19:21:32  GOPATH: 
2015/08/20 19:21:32  GOOS: linux
2015/08/20 19:21:32  GOARCH: amd64
2015/08/20 19:21:32  lib-path: ""
2015/08/20 19:21:32 extracted expression tokens: []main.token_item{main.token_item{off:41, tok:4, lit:"b"}}
2015/08/20 19:21:32 Offset: 0
2015/08/20 19:21:32 Number of candidates found: 0
2015/08/20 19:21:32 Candidates are:
2015/08/20 19:21:32 =======================================================
```

If I run the client like this:

    $ GOPATH=${GB_PROJECT_DIR}:${GB_PROJECT_DIR}/vendor gocode -in src/a/a.go autocomplete 43

the result is the same (the GOPATH line changes accordingly). I have tried both GOPATH=${GB_PROJECT_DIR}:${GB_PROJECT_DIR}/vendor and GOPATH=${GB_PROJECT_DIR}/vendor:${GB_PROJECT_DIR}.

Digging in gocode, the code is hitting this:

https://github.com/nsf/gocode/blob/master/declcache.go#L256-L312

It's not entering this:

	if g_config.LibPath != "" {
		for _, p := range filepath.SplitList(g_config.LibPath) {
			pkg_path := filepath.Join(p, pkgfile)
			if file_exists(pkg_path) {
				log_found_package_maybe(imp, pkg_path)
				return pkg_path, true
			}
			// Also check the relevant pkg/OS_ARCH dir for the libpath, if provided.
			pkgdir := fmt.Sprintf("%s_%s", context.GOOS, context.GOARCH)
			pkg_path = filepath.Join(p, "pkg", pkgdir, pkgfile)
			if file_exists(pkg_path) {
				log_found_package_maybe(imp, pkg_path)
				return pkg_path, true
			}
			// Also check the relevant pkg/OS/ARCH dir for the libpath, if provided.
			pkg_path = filepath.Join(p, "pkg", context.GOOS, context.GOARCH, pkgfile)
			if file_exists(pkg_path) {
				log_found_package_maybe(imp, pkg_path)
				return pkg_path, true
			}
		}
	}

because the corresponding configuration option (lib-path) is not set.

It's hitting this:

	p, err := context.Import(imp, "", build.AllowBinary|build.FindOnly)

here context is a build.Context from go/build. The return value in p is:

    build.Package{
        Dir:"/home/magallma/devel/go/test/vendor/src/b",
        Name:"", ImportComment:"",
        Doc:"", ImportPath:"b",
        Root:"/home/magallma/devel/go/test/vendor",
        SrcRoot:"/home/magallma/devel/go/test/vendor/src",
        PkgRoot:"/home/magallma/devel/go/test/vendor/pkg",
        PkgTargetRoot:"/home/magallma/devel/go/test/vendor/pkg/linux_amd64",
        BinDir:"/home/magallma/devel/go/test/vendor/bin",
        Goroot:false,
        PkgObj:"/home/magallma/devel/go/test/vendor/pkg/linux_amd64/b.a",
        AllTags:[]string(nil),
        ConflictDir:"",
        GoFiles:[]string(nil),
        CgoFiles:[]string(nil),
        IgnoredGoFiles:[]string(nil),
        CFiles:[]string(nil),
        CXXFiles:[]string(nil),
        MFiles:[]string(nil),
        HFiles:[]string(nil),
        SFiles:[]string(nil),
        SwigFiles:[]string(nil),
        SwigCXXFiles:[]string(nil),
        SysoFiles:[]string(nil),
        CgoCFLAGS:[]string(nil),
        CgoCPPFLAGS:[]string(nil),
        CgoCXXFLAGS:[]string(nil),
        CgoLDFLAGS:[]string(nil),
        CgoPkgConfig:[]string(nil),
        Imports:[]string(nil),
        ImportPos:map[string][]token.Position(nil),
        TestGoFiles:[]string(nil),
        TestImports:[]string(nil),
        TestImportPos:map[string][]token.Position(nil),
        XTestGoFiles:[]string(nil),
        XTestImports:[]string(nil),
        XTestImportPos:map[string][]token.Position(nil)
    }

notice how context.PkgObj is "/home/magallma/devel/go/test/vendor/pkg/linux_amd64/b.a", a path inside *vendor*.

I went looking at https://github.com/golang/go/blob/master/src/go/build/build.go#L459 and that method is finding the source (around line 563) in vendor/src/b/b.go and it then compues the PkgObj as shown above around https://github.com/golang/go/blob/master/src/go/build/build.go#L599.

Recently gocode was patched to deal with the $GOOS/$GOARCH vs ${GOOS}_${GOARCH} difference between the go tool and gb (https://github.com/nsf/gocode/commit/85c63bdec7cd2dc6531cf2cb52d243a914ce6744), but this is a different issue.

What I'm thinking is that tools that wish to support gb would need some help from gb in the sense that they need a way to figure out where binaries are landing, which is something that gb is already computing internally. They would need to import a tiny (?) bit of code from gb in order to pass a package import path and gb would return something akin to go/build's Context. Does this sound like a good avenue to pursue?

Last but not least, this problem can be fixed by setting lib-path to ${GB_PROJECT_DIR}, but as far as I can tell this is a global setting, which doesn't bode well with a per-project pkg directory.

cc: @MerlinDMC

**Comment by mem** [comment]
I should note this here as well: the issue manifests itself with gocode built with Go 1.5, but not with gocode built with Go 1.4. I haven't narrowed down the change in Go 1.5 that causes this.

**Comment by nsf** [comment]
So, ugm, but do you actually build "b"? Gocode works only with binary packages. From what I've seen above, context.Import actually finds the package, so what's the problem?

**Comment by nsf** [comment]
I mean if context.Import returns the right path, the only place later where gocode fails is this line: https://github.com/nsf/gocode/blob/master/declcache.go#L300

Unless "autobuild" feature is turned on. So, yeah, binary package, is it actually there?

**Comment by nsf** [comment]
Or maybe the path is wrong, sorry, this vendoring thing is very confusing to me. :D I should read the docs first.

**Comment by nsf** [comment]
@mem may I ask where GB_PROJECT_DIR comes from? I see gb uses it internally somewhat. But what makes you think it is set to something externally and globally?

**Comment by davecheney** [comment]
It's the root of the gb project. So whenever people talk about the gb
project, or use the form $PROJECT/src, that's GB_PROJECT_DIR. It's also
inserted into the environment of any gb-plugin.

**Comment by nsf** [comment]
@davecheney Yeah, I understand, but @mem says he does: `GOPATH=${GB_PROJECT_DIR}:${GB_PROJECT_DIR}/vendor gocode -in src/a/a.go autocomplete 43` and it doesn't help. Which looks weird to me. Like is it the actual command he uses? Why would GB_PROJECT_DIR be set in his shell environment. The only debug output from gocode I see is without GOPATH set. So what if GB_PROJECT_DIR isn't even set to anything. I mean it could be as well something like this:

```
[nsf @ ~]$ export GOPATH=${GB_PROJECT_DIR}:${GB_PROJECT_DIR}/vendor
[nsf @ ~]$ echo $GOPATH
:/vendor
```

**Comment by davecheney** [comment]
I don't believe if would be.

**Comment by davecheney** [comment]
With that said, gb offers the info command which can synthesise these
details
```
% gb info
GB_PROJECT_DIR="/home/dfc"
GB_SRC_PATH="/home/dfc/src:/home/dfc/vendor/src"
```

**Comment by nsf** [comment]
@mem can you please show me gocode's debug output when you do:
```
GOPATH=${GB_PROJECT_DIR}:${GB_PROJECT_DIR}/vendor gocode -in src/a/a.go autocomplete 43
```

**Comment by mem** [comment]
I have:

```
$ find ! -type d
./src/a/a.go
./vendor/src/b/b.go

$ cat ./src/a/a.go 
package a

import "b"

func A() {
        b.B()
}

$ cat ./vendor/src/b/b.go
package b

func B() {
}
```

I can build:

```
$ gb build
b
a
$ find ! -type d
./pkg/linux/amd64/b.a
./pkg/linux/amd64/a.a
./src/a/a.go
./vendor/src/b/b.go
```

I can start gocode and send an autocompletion request:

The request:
```
$ GOPATH=$PWD:$PWD/vendor gocode -in src/a/a.go autocomplete 37
Nothing to complete.
```

The server:
```
$ gocode -s -sock unix -addr localhost:37373 -debug
2015/08/22 08:14:34 Got autocompletion request for '/home/magallma/devel/go/test/src/a/a.go'
2015/08/22 08:14:34 Cursor at: 37
2015/08/22 08:14:34 -------------------------------------------------------
package a

import "b"

func A() {
        b.#B()
}
2015/08/22 08:14:34 -------------------------------------------------------
2015/08/22 08:14:34 Import path "b" was not resolved
2015/08/22 08:14:34 Gocode's build context is:
2015/08/22 08:14:34  GOROOT: /home/magallma/devel/go/go-1.5
2015/08/22 08:14:34  GOPATH: /home/magallma/devel/go/test:/home/magallma/devel/go/test/vendor
2015/08/22 08:14:34  GOOS: linux
2015/08/22 08:14:34  GOARCH: amd64
2015/08/22 08:14:34  lib-path: ""
2015/08/22 08:14:34 extracted expression tokens: []main.token_item{main.token_item{off:35, tok:4, lit:"b"}}
2015/08/22 08:14:34 Offset: 0
2015/08/22 08:14:34 Number of candidates found: 0
2015/08/22 08:14:34 Candidates are:
2015/08/22 08:14:34 =======================================================
```

I can do this:

```
$ gocode set lib-path $PWD:$PWD/vendor
```

and reissue the request, and then I get:

```
2015/08/22 08:16:57 Got autocompletion request for '/home/magallma/devel/go/test/src/a/a.go'
2015/08/22 08:16:57 Cursor at: 37
2015/08/22 08:16:57 -------------------------------------------------------
package a

import "b"

func A() {
        b.#B()
}
2015/08/22 08:16:57 -------------------------------------------------------
2015/08/22 08:16:57 Found "b" at "/home/magallma/devel/go/test/pkg/linux/amd64/b.a"
2015/08/22 08:16:57 extracted expression tokens: []main.token_item{main.token_item{off:35, tok:4, lit:"b"}}
2015/08/22 08:16:57 Offset: 0
2015/08/22 08:16:57 Number of candidates found: 1
2015/08/22 08:16:57 Candidates are:
2015/08/22 08:16:57   func B()
2015/08/22 08:16:57 =======================================================
```

**Comment by nsf** [comment]
Well, that's because gb uses `$GOOS/$GOARCH` path instead of `$GOOS_$GOARCH`. And this patch (https://github.com/nsf/gocode/commit/85c63bdec7cd2dc6531cf2cb52d243a914ce6744) was meant to be used with lib-path.

So I don't see what exactly is wrong. Because GOPATH doesn't have to work with gb's directory scheme. And the gb support is here via lib-path.

**Comment by mem** [comment]
@nsf I'm sorry if I'm making this harder to understand than absolutely needed...

Like I wrote before (first comment):

"Recently gocode was patched to deal with the $GOOS/$GOARCH vs ${GOOS}_${GOARCH} difference between the go tool and gb (nsf/gocode@85c63bd), but this is a different issue.

Last but not least, this problem can be fixed by setting lib-path to ${GB_PROJECT_DIR}, but as far as I can tell this is a global setting, which doesn't bode well with a per-project pkg directory."

Also in that first message, the call to context.Import is returning:
```
    SrcRoot:"/home/magallma/devel/go/test/vendor/src",
    PkgRoot:"/home/magallma/devel/go/test/vendor/pkg",
```
note that it says `vendor/src` and `vendor/pkg`. `gb build` on the other hand puts the binaries in `pkg` (no "vendor"), so even with the other issue (/ vs _) out of the way, this would not work.

Also, note that gb's directory scheme _is_ compatible with the way the go tool works. I can go to a clean gb project directory and do this:
```
GOPATH=$PWD:$PWD/vendor go install ./...
```
and this builds all of the packages in the project, with the caveat that packages under `vendor/src` have their binaries installed to `vendor/pkg` and `vendor/bin`. godef does work in this case if GOPATH is set as shown above.

**Comment by nsf** [comment]
Yes, `lib-path` is a global setting. But gocode never really was designed to work well with multiple users simultaneously. So, setting `lib-path` before each request is okay. The point here that your editor needs to do that anyway. Gocode have no idea what a "project dir" is. Your editor might. And if editor also knows about gb, it will do whatever it feels like, be it setting `lib-path` or for that matter setting $GOPATH, even though it's a wrong thing to do. Gocode provides gb dir structure support only via `lib-path`, because $GOPATH is a standard go tool thing, not a gb thing.

I disagree with you that `GOPATH=$PWD:$PWD/vendor go install ./...` is the same as using gb. Because the packages will be installed into a different directory. Go tool uses `$GOPATH/pkg/$GOOS_$GOARCH` dir. Gb uses `$PROJECT/pkg/$GOOS/$GOARCH`. See, I've executed both `gb build` and `go install` here:

```
[nsf @ libtest]$ tree
.
├── pkg
│   ├── linux
│   │   └── amd64
│   │       └── libtest.a
│   └── linux_amd64
│       └── libtest.a
└── src
    └── libtest
        └── test.go

6 directories, 3 files
```

Btw, gb author, why??? Is it hard to put everything under the same dir scheme go tool does to make things compatible with tools? Well, let's make gb users suffer then.


Stop trying to make gocode work with gb via $GOPATH.

If you want vendoring, I will eventually add vendoring support go tool provides. Which is experimental right now and it's not worth adding, because they may change it. As for gb support in gocode, it's possible to add the same project dir discovery mechanics into gocode, but I will not do so, unless gb proves to be a de-facto standard. All I can offer is `lib-path` and it was accepted in that patch. In fact it was an exception from my side, because I simply cannot support every go project management tool out there. And even if I'd be okay with that, adding every feature means maintenance burden, I have no time to do that maintenance and tools may decide to change their mechanics at arbitrary points in time.

**Comment by dbehnke** [comment]
For anyone encountering this issue, I've had success working around this problem by using the suggested lib-path of gocode.  I create and activate an environment via  source ./activate_env (similar to how python does it for virtual environments) -- example activate: https://github.com/dbehnke/golang-gb-project-template/blob/master/activate_env   Then I issue a "gocode close" to make sure no gocode daemons are running.  Finally, I launch vim or atom or whatever editor I'm using from the terminal to ensure the environment is being used properly.    Code completion works on compiled packages this way.   

If you are writing source however, code completion doesn't work until it's successfully built... (I'm not sure if this is how gocode works by default - I'm making an assumption here that is does or there would be no need for that auto build flag).  So effectively you are able to get code completion on the last successful compile.   

Both gb and gocode are great tools in themselves.  Hopefully once Go's vendoring experiment is no longer experimental.. this issue will resolve itself on its own.

**Comment by nsf** [comment]
Also guys, I'm working on a proprietary project right now which uses godep internally and I'm super unhappy about it and in general with the whole vendoring story. I'm looking into gb right now and see if it can solve my problems. If it can, I will definitely consider adding it to gocode, just because I need it and if I think it's the right way to solve the longstanding problem. For a long time I was ignoring the whole vendoring issue, but now it affects me and what I do directly.

**Comment by nsf** [comment]
Added basic gb-specific lookup to gocode, it is enabled via new "package-lookup-mode" option, see README.md for details. But I'm not sure where I'll end up with this feature. 

What I like about gb is the fact that it doesn't mess with your GOPATH. It is absolutely great. And I also like that gb makes it clear - vendoring in libraries is a bad idea. But it also means that you'll end up developing your libraries under your GOPATH often. Technically you can use gb there too and perhaps you should. But if you want to use go tool instead, gocode's "package-lookup-mode" will require you to switch it often.

Maybe gb will be able to fully replace go tool one day? Right now if I understand correctly even though it has features like gb-vendor, it can't replace "go get". But since it's a plugin-based architecture, no one prevents us from building gb-get plugin. :)

As I mentioned in the previous comment, I had issues with godep on the project I'm working on and gb seems to work better for us.
